fixham/internal/release/releasenotes.go

```diff
@@ -0,0 +1,228 @@
+package release
+
+import (
+	"context"
+	"fmt"
+	"os"
+	"path/filepath"
+	"strconv"
+	"strings"
+	"text/template"
+	"time"
+
+	"github.com/google/go-github/v53/github"
+	"github.com/sirupsen/logrus"
+
+	"github.com/projectcalico/calico/fixham/internal/utils"
+	"github.com/projectcalico/calico/fixham/internal/version"
+)
+
+const (
+	releaseNoteLabel        = "release-note-required"
+	releaseNotesFolder      = "release-notes"
+	releaseNoteTemplatePath = "/fixham/assets/release-note.md.tmpl"
+)
+
+var (
+	repos = []string{"calico", "bird"}
+)
+
+// ReleaseNoteIssueData represents the data for an release note issue
+type ReleaseNoteIssueData struct {
+	ID     int
+	Note   string
+	Repo   string
+	URL    string
+	Author string
+}
+
+// ReleaseNoteData represents the data for release notes
+type ReleaseNoteData struct {
+	Date         string
+	BugFixes     []*ReleaseNoteIssueData
+	OtherChanges []*ReleaseNoteIssueData
+}
+
+// milestoneNumber returns the milestone number for a given milestone
+func milestoneNumber(client *github.Client, owner, repo, milestone string) (int, error) {
+	for {
+		milestones, resp, err := client.Issues.ListMilestones(context.Background(), owner, repo, nil)
+		if err != nil {
+			return -1, err
+		}
+		for _, m := range milestones {
+			if m.GetTitle() == milestone {
+				return m.GetNumber(), nil
+			}
+		}
+		if resp.NextPage == 0 {
+			break
+		}
+	}
+	return -1, fmt.Errorf("milestone not found")
+}
+
+// prIssuesByRepo returns all the PR issues for a given repo
+func prIssuesByRepo(client *github.Client, owner, repo string, opts *github.IssueListByRepoOptions) ([]*github.Issue, error) {
+	prIssues := []*github.Issue{}
+	for {
+		issues, resp, err := client.Issues.ListByRepo(context.Background(), owner, repo, opts)
+		if err != nil {
+			return nil, err
+		}
+		for _, issue := range issues {
+			if issue.IsPullRequest() {
+				prIssues = append(prIssues, issue)
+			}
+		}
+		if resp.NextPage == 0 {
+			break
+		}
+		opts.Page = resp.NextPage
+	}
+	return prIssues, nil
+}
+
+// extractReleaseNoteFromIssue extracts release notes from an issue.
+// It looks for the release note block in the issue body and returns the content
+// between the start and end markers.
+func extractReleaseNoteFromIssue(issue *github.Issue) ([]string, error) {
+	body := issue.GetBody()
+	startMarker := "```release-note"
+	endMarker := "```"
+	startIndex := strings.Index(body, startMarker)
+	if startIndex == -1 {
+		return []string{issue.GetTitle()}, fmt.Errorf("start marker not found")
+	}
+	startIndex += len(startMarker)
+	endIndex := strings.Index(body[startIndex:], endMarker)
+	if endIndex == -1 {
+		return nil, fmt.Errorf("end marker not found")
+	}
+	notes := strings.TrimSpace(body[startIndex : startIndex+endIndex])
+	if len(notes) == 0 {
+		return []string{issue.GetTitle()}, fmt.Errorf("no release notes found")
+	}
+	return strings.Split(notes, "\n"), nil
```

Comment:
it's possible for a single release note to span multiple lines, so I don't think we want to split them.